source/connection_manager.c

```diff
@@ -310,14 +312,43 @@ static bool s_aws_http_connection_manager_should_destroy(struct aws_http_connect
  * simply by setting the error_code and moving it to the current transaction's completion list.
  */
 struct aws_http_connection_acquisition {
+    struct aws_allocator *allocator;
     struct aws_linked_list_node node;
     struct aws_http_connection_manager *manager; /* Only used by logging */
     aws_http_connection_manager_on_connection_setup_fn *callback;
     void *user_data;
     struct aws_http_connection *connection;
     int error_code;
+    struct aws_channel_task acquisition_task;
 };
 
+static void s_connection_acquisition_task(
+    struct aws_channel_task *channel_task,
+    void *arg,
+    enum aws_task_status status) {
+    (void)channel_task;
+
+    struct aws_http_connection_acquisition *pending_acquisition = arg;
+
+    if (status != AWS_TASK_STATUS_RUN_READY) {
+        AWS_LOGF_WARN(
+            AWS_LS_HTTP_CONNECTION_MANAGER,
+            "id=%p: Failed to completed connection acquisition because the connection was closed",
+            (void *)pending_acquisition->manager);
+        pending_acquisition->callback(NULL, AWS_ERROR_HTTP_CONNECTION_CLOSED, pending_acquisition->user_data);
```

Comment:
What are the conditions when this would happen?  I can only see an event loop shutdown being the trigger and if that's happening (while the connection manager is still alive and it is alive because this is a vended connection at this point, preventing connection manager shutdown), then we're in a world of hurt.  

I'm not able to see the link between the task being in a cancelled state and the connection being closed.

I am also worried that this leaks the connection.  The connection was good before the task was scheduled and the connection manager has already incremented its vended count at that time.  If we don't actually end up vending the connection then we need to perform the proper fix up here and possibly release the connection back to the pool.

Reply:
It’s a channel task though

Reply:
This would only happen if the channel was shutdown, which simply means, the connection closed on you. These are channel tasks, not event-loop tasks.

Good catch on releasing the connection back to the pool though.